### DISCUSSION TITLE
Bitwuzla performance

### DISCUSSION THREAD

**Discussion by apach301** [bitwuzla/bitwuzla#100] [General]
Hi!

I'm working on symbolic execution tool Sydr, which uses Bitwuzla as a solver backend. It use pretty old 1.0-prerelease version (`f834fc4909c232e695c036c42eb98f9ac26564d1`) via Triton framework (https://github.com/JonathanSalwan/Triton).

But I found out that updating Bitwuzla to the newest version causes a huge performance regression. Triton uses Bitwuzla via C (C++ in the new version) API, so I tried to build and test Bitwuzla as a binary on a set of SMT-LIB2 queries dumped to files ([query_set.tar.gz](https://github.com/bitwuzla/bitwuzla/files/13965026/query_set.tar.gz)).

To check Bitwuzla performance I simply run `bitwuzla` on the set of queries and measure the common execution time: `time for q in query_set/*; do bitwuzla_install_dir/bin/bitwuzla $q; done`. I get next results: release 0.3.0 process all set for 16.8s, release 0.2.0 for ~12.5s, and old `f834fc4`-version for ~2.1s.

What could cause such performance drop and how to eliminate this? Maybe I missed some settings/optimizations/flags or using Bitwuzla in a wrong way? I'm building Bitwuzla with `./configure.py release --prefix $PWD/build/bzla_install --no-testing --no-unit-testing && cd build && meson compile && meson install`. Setup: Ubuntu 20.04 amd64, Python 3.8. Missing dependencies are fetched by configure command:
![image](https://github.com/bitwuzla/bitwuzla/assets/13482580/56c526c3-f38d-4281-bb76-4ee8c002aeba)

**Comment by mpreiner** [comment]
Thanks for the benchmarks, we'll look into this!

**Reply by recoules** [reply]
Hi @apach301, @mpreiner,

As an educated guess, I would first try to rerun the bench with the `PP_SKELETON_PREPROC` option disabled.
I experienced a similar slowdown by upgrading `bitwuzla` and found that this preprocessing can be quite expensive for simple but numerous queries.

Regards

**Reply by apach301** [reply]
Hi, thanks for reply! Tried to rerun benchmark with:

- `bitwuzla --pp-skeleton-preproc=false $file` -> ~12.0s
- `bitwuzla --pp-skeleton-preproc=true $file` -> ~12.2s

So this option have a very little effect. Maybe there are some other options that changed by default since old commits.

**Reply by recoules** [reply]
Ok, sorry, it would have been an "easy" answer.

**Comment by apach301** [comment]
I also tried to test different commits, but some of them are not easy to build. For now I discovered that the latest Bitwuzla version that have no performance issues (compared to "base" `f834fc4`) - is `fefaa8c16019598ce00e9ab76fe41727dbcd4ddc`. Starting from `ac4b4806cf182f636efc057ff1168eb26fd41fb8` and up to the HEAD it is already very slow (~12.5s on query_set). Therefore, changes that affect performance should be somewhere between these two commits, but I failed to build Bitwuzla on these commits as it missing `bitwuzla-libs` dependency.

**Reply by mpreiner** [reply]
fefaa8c16019598ce00e9ab76fe41727dbcd4ddc is still the old code base, between this one and main was a complete rewrite of the solver :) I'm currently extremely busy due to deadlines, but I'll have a look at this issue starting next week.

**Comment by mpreiner** [comment]
Alright some updates. There was a performance issue in the SMT2  parser, which is now fixed with 9b46ba9ba2cc6704f6c2368bc8c017def33884d7 and only occurred when no logic was set via the `set-logic` command. This should cut the runtime in half, but it is still more than 2x slower than the old code base. I have a few ideas to improve this, but this will take more time. I'll keep you posted on the progress.

**Reply by apach301** [reply]
Thanks for the fix! I checked this out - it really helps to improve the speed (from 12s to 8s). I also found a bug in my build (now I set CC/CXX to clang toolchain) - as a result the build is now running 4s. But it still 2 times slower than the old version (2s), so it's great if there are more options to speed it up.

**Comment by mpreiner** [comment]
I pushed some more improvements, which should be again about 2x faster than the previous version. Let me know how that works for you.

**Reply by apach301** [reply]
For my build the time had reduced to ~3.0s on that mini-benchmark (while the old version works for ~2.15s). In addition, I checked performance on more formulas produced from different test-programs in our symbolic benchmark. Now fixed bitwuzla works ~1.5x times slower than old version - which is much better than bitwuzla v0.3.2. I guess while it's still slower than old version (tried to build with different gmp/cadical releases but it had no effect). Here's larger benchmark I performed tests on: [benchmark_predicates.zip](https://github.com/bitwuzla/bitwuzla/files/14209196/benchmark_predicates.zip)

**Reply by apach301** [reply]
Btw, when I launch bitwuzla it always ends with segfault, so I this patch (https://github.com/bitwuzla/bitwuzla/issues/96#issuecomment-1851895397) for all my local builds. It would be great if this fix would be merged into main.

**Reply by mpreiner** [reply]
I've added the workaround on main.
